## 1. Series로 feature를 보다 상세하게 탐색하기

#### **1.1** 코로나 바이러스 데이터와 함께 pandas 라이브러리 익히기
- COVID-19-master 폴더 확인

In [4]:
import pandas as pd
PATH = "00_Material(Uploaded)/COVID-19-master/csse_covid_19_data/csse_covid_19_daily_reports/"
doc = pd.read_csv(PATH + "04-01-2020.csv", encoding='utf-8-sig')

In [6]:
doc.head()

FIPS     Admin2  Province_State Country_Region          Last_Update  \
0  45001.0  Abbeville  South Carolina             US  2020-04-01 21:58:49   
1  22001.0     Acadia       Louisiana             US  2020-04-01 21:58:49   
2  51001.0   Accomack        Virginia             US  2020-04-01 21:58:49   
3  16001.0        Ada           Idaho             US  2020-04-01 21:58:49   
4  19001.0      Adair            Iowa             US  2020-04-01 21:58:49   

         Lat       Long_  Confirmed  Deaths  Recovered  Active  \
0  34.223334  -82.461707          4       0          0       4   
1  30.295065  -92.414197         47       1          0      46   
2  37.767072  -75.632346          7       0          0       7   
3  43.452658 -116.241552        195       3          0     192   
4  41.330756  -94.471059          1       0          0       1   

                    Combined_Key  
0  Abbeville, South Carolina, US  
1          Acadia, Louisiana, US  
2         Accomack, Virginia, US  
3                 Ada, Idaho, US  
4                Adair, Iowa, US

#### **1.2** 데이터프레임에서 Series 추출하기
- 하나의 feature(column)만 선택하면 됨

In [15]:
countries = doc['Country_Region']
countries.head() # countries만 해서 전체를 봐도 되지만 데이터가 너무 많기 때문에 보통 .head()를 사용함

0    US
1    US
2    US
3    US
4    US
Name: Country_Region, dtype: object

#### **1.3** Series로 feature를 보다 상세하게 탐색하기
- `size` : 사이즈 반환
- `count()` : 데이터가 없는 경우를 뺀 사이즈 반환
- `unique()`: 유일한 값만 반환
- `value_counts()`: 데이터가 없는 경우를 제외하고, 각 값의 갯수를 반환

In [18]:
print(countries.size, countries.count())

2522 2522


In [20]:
print (countries.unique(), len(countries.unique()))

['US' 'Canada' 'United Kingdom' 'China' 'Netherlands' 'Australia'
 'Denmark' 'France' 'Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Austria' 'Azerbaijan'
 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus' 'Belgium' 'Belize'
 'Benin' 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina' 'Botswana' 'Brazil'
 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burma' 'Burundi' 'Cabo Verde'
 'Cambodia' 'Cameroon' 'Central African Republic' 'Chad' 'Chile'
 'Colombia' 'Congo (Brazzaville)' 'Congo (Kinshasa)' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus' 'Czechia' 'Diamond Princess'
 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Eswatini'
 'Ethiopia' 'Fiji' 'Finland' 'Gabon' 'Gambia' 'Georgia' 'Germany' 'Ghana'
 'Greece' 'Grenada' 'Guatemala' 'Guinea' 'Guinea-Bissau' 'Guyana' 'Haiti'
 'Holy See' 'Honduras' 'Hungary' 'Iceland' 'India' 'Indonesia' 'Iran'
 'Iraq' 'Ireland' 'Israel' 'It

In [22]:
countries.value_counts()

Country_Region
US                2228
China               34
United Kingdom      18
Malaysia            17
Canada              16
                  ... 
Ghana                1
Greece               1
Grenada              1
Guatemala            1
Tuvalu               1
Name: count, Length: 190, dtype: int64

> data는 총 2522개가 있고, 유니크한 데이터는 190개이고, 미국 정보가 매우 많다는 것을 알 수 있음

## 2. 필요한 컬럼만 선택하기
- 여러 컬럼을 선택하면, 별도의 데이터프레임이 됨

In [26]:
covid_stat = doc[['Confirmed', 'Deaths', 'Recovered']]
covid_stat.head()

Confirmed  Deaths  Recovered
0          4       0          0
1         47       1          0
2          7       0          0
3        195       3          0
4          1       0          0

## 3. 특정 조건에 맞는 row 검색하기
- 미국 데이터만 따로 보고 싶다면?

In [31]:
doc_us = doc[doc['Country_Region'] == 'US']
doc_us

FIPS     Admin2            Province_State Country_Region  \
0     45001.0  Abbeville            South Carolina             US   
1     22001.0     Acadia                 Louisiana             US   
2     51001.0   Accomack                  Virginia             US   
3     16001.0        Ada                     Idaho             US   
4     19001.0      Adair                      Iowa             US   
...       ...        ...                       ...            ...   
2246  66000.0        NaN                      Guam             US   
2273      NaN        NaN  Northern Mariana Islands             US   
2280      NaN        NaN               Puerto Rico             US   
2285      NaN        NaN                 Recovered             US   
2303      NaN        NaN            Virgin Islands             US   

              Last_Update        Lat       Long_  Confirmed  Deaths  \
0     2020-04-01 21:58:49  34.223334  -82.461707          4       0   
1     2020-04-01 21:58:49  30.295065  -92.414197         47       1   
2     2020-04-01 21:58:49  37.767072  -75.632346          7       0   
3     2020-04-01 21:58:49  43.452658 -116.241552        195       3   
4     2020-04-01 21:58:49  41.330756  -94.471059          1       0   
...                   ...        ...         ...        ...     ...   
2246  2020-04-01 21:58:49  13.444300  144.793700         77       3   
2273  2020-04-01 21:58:49  15.097900  145.673900          2       0   
2280  2020-04-01 21:58:49  18.220800  -66.590100        286      11   
2285  2020-04-01 21:58:49   0.000000    0.000000          0       0   
2303  2020-04-01 21:58:49  18.335800  -64.896300         30       0   

      Recovered  Active                   Combined_Key  
0             0       4  Abbeville, South Carolina, US  
1             0      46          Acadia, Louisiana, US  
2             0       7         Accomack, Virginia, US  
3             0     192                 Ada, Idaho, US  
4             0       1                Adair, Iowa, US  
...         ...     ...                            ...  
2246          0      74                       Guam, US  
2273          0       2   ,Northern Mariana Islands,US  
2280          0     275                Puerto Rico, US  
2285       8474       0                  Recovered, US  
2303          0      30             ,Virgin Islands,US  

[2228 rows x 12 columns]

## 4. 없는 데이터(NaN) 처리하기

#### **4.1** 없는 데이터(결측치)가 있는지 확인하기
- isnull() : 없는 데이터가 있는지 확인 (True or False)
- sum() : 없는 데이터가 있는 행의 갯수 확인
- 통상 isnull().sum() 으로 사용

In [38]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc.isnull().sum()

Province/State    12
Country/Region     0
Last Update        0
Confirmed         10
Deaths            37
Recovered         37
dtype: int64

> Province/State 컬럼에는 결측치가 12개 있다는 것

#### **4.2** 없는 데이터 삭제하기
- `dropna()` : 결측치를 가진 행을 모두 삭제

In [48]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc.dropna()
doc.isnull().sum()

Province/State    0
Country/Region    0
Last Update       0
Confirmed         0
Deaths            0
Recovered         0
dtype: int64

#### **4.3** 특정 컬럼 값이 없는 데이터만 삭제하기
- 데이터를 많이 삭제하지 않을 수 있음
- `subset`으로 해당 컬럼을 지정해줌

In [51]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc.dropna(subset=['Confirmed'])
doc.head()

Province/State  Country/Region      Last Update  Confirmed  Deaths  \
0          Anhui  Mainland China  1/22/2020 17:00        1.0     NaN   
1        Beijing  Mainland China  1/22/2020 17:00       14.0     NaN   
2      Chongqing  Mainland China  1/22/2020 17:00        6.0     NaN   
3         Fujian  Mainland China  1/22/2020 17:00        1.0     NaN   
5      Guangdong  Mainland China  1/22/2020 17:00       26.0     NaN   

   Recovered  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
5        NaN

#### **4.4** 없는 데이터(NaN)을 특정값으로 일괄 변경하기
- `fillna(특정값)` : 특정값으로 결측치를 대체

In [54]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc.fillna(0) # NaN 값을 0으로 체체
doc.head()

Province/State  Country/Region      Last Update  Confirmed  Deaths  \
0          Anhui  Mainland China  1/22/2020 17:00        1.0     0.0   
1        Beijing  Mainland China  1/22/2020 17:00       14.0     0.0   
2      Chongqing  Mainland China  1/22/2020 17:00        6.0     0.0   
3         Fujian  Mainland China  1/22/2020 17:00        1.0     0.0   
4          Gansu  Mainland China  1/22/2020 17:00        0.0     0.0   

   Recovered  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0

#### **4.5** 없는 데이터(NaN) 중 특정 컬럼에 대해 특정 값으로 일괄 변경하기
- 별도 사전 데이터를 생성하고
- 없는 데이터를 변경할 `컬럼명`만 `키`로 만들고
- `변경할 특정 값`을 `키값`으로 넣고
- `fillna()` 함수에 적용해주면 됨

In [57]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
nan_data = {'Deaths': 0, 'Recovered':0}
doc = doc.fillna(nan_data)
doc.head()

Province/State  Country/Region      Last Update  Confirmed  Deaths  \
0          Anhui  Mainland China  1/22/2020 17:00        1.0     0.0   
1        Beijing  Mainland China  1/22/2020 17:00       14.0     0.0   
2      Chongqing  Mainland China  1/22/2020 17:00        6.0     0.0   
3         Fujian  Mainland China  1/22/2020 17:00        1.0     0.0   
4          Gansu  Mainland China  1/22/2020 17:00        NaN     0.0   

   Recovered  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0

> 만약 Country/Region이 NaN일 때 doc.fillna(0)을 하면 Country/Region 값이 0이 됨. 이런 이상한 값을 방지하려고 하는 것

## 5. 특정 키값을 기준으로 데이터 합치기
- `groupby()` : SQL 구문의 group by 와 동일, 특정 컬럼을 기준으로 그룹
- `sum()` : 그룹으로 되어 있는 데이터를 합치기

In [61]:
doc = pd.read_csv(PATH + "04-01-2020.csv", encoding='utf-8-sig')
doc.head()

FIPS     Admin2  Province_State Country_Region          Last_Update  \
0  45001.0  Abbeville  South Carolina             US  2020-04-01 21:58:49   
1  22001.0     Acadia       Louisiana             US  2020-04-01 21:58:49   
2  51001.0   Accomack        Virginia             US  2020-04-01 21:58:49   
3  16001.0        Ada           Idaho             US  2020-04-01 21:58:49   
4  19001.0      Adair            Iowa             US  2020-04-01 21:58:49   

         Lat       Long_  Confirmed  Deaths  Recovered  Active  \
0  34.223334  -82.461707          4       0          0       4   
1  30.295065  -92.414197         47       1          0      46   
2  37.767072  -75.632346          7       0          0       7   
3  43.452658 -116.241552        195       3          0     192   
4  41.330756  -94.471059          1       0          0       1   

                    Combined_Key  
0  Abbeville, South Carolina, US  
1          Acadia, Louisiana, US  
2         Accomack, Virginia, US  
3                 Ada, Idaho, US  
4                Adair, Iowa, US

In [65]:
# confirmed를 볼 것이기 때문에 sum() 사용해서 각 국가마다 확진자 수 합치기
doc = doc.groupby('Country_Region').sum() 
doc.head()

FIPS Admin2 Province_State          Last_Update       Lat  \
Country_Region                                                              
Afghanistan      0.0      0              0  2020-04-01 21:58:34  33.93911   
Albania          0.0      0              0  2020-04-01 21:58:34  41.15330   
Algeria          0.0      0              0  2020-04-01 21:58:34  28.03390   
Andorra          0.0      0              0  2020-04-01 21:58:34  42.50630   
Angola           0.0      0              0  2020-04-01 21:58:34 -11.20270   

                    Long_  Confirmed  Deaths  Recovered  Active Combined_Key  
Country_Region                                                                
Afghanistan     67.709953        192       4          5     183  Afghanistan  
Albania         20.168300        259      15         67     177      Albania  
Algeria          1.659600        847      58         61     728      Algeria  
Andorra          1.521800        390      14         10     366      Andorra  
Angola          17.873900          8       2          1       5       Angola

#### **5.1** groupby 에 의해서, index가 Country_Region 의 각 국가로 변경됨

In [68]:
doc.columns

Index(['FIPS', 'Admin2', 'Province_State', 'Last_Update', 'Lat', 'Long_',
       'Confirmed', 'Deaths', 'Recovered', 'Active', 'Combined_Key'],
      dtype='object')

In [70]:
doc.index

Index(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Antarctica',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       ...
       'United Arab Emirates', 'United Kingdom', 'Uruguay', 'Uzbekistan',
       'Venezuela', 'Vietnam', 'West Bank and Gaza', 'Winter Olympics 2022',
       'Zambia', 'Zimbabwe'],
      dtype='object', name='Country_Region', length=190)

#### **5.2** index로 검색해서 US 의 합계 검색 가능

In [74]:
doc[doc.index == 'US']

FIPS                                             Admin2  \
Country_Region                                                                  
US              65168934.0  AbbevilleAcadiaAccomackAdaAdairAdairAdairAdams...   

                                                   Province_State  \
Country_Region                                                      
US              South CarolinaLouisianaVirginiaIdahoIowaMissou...   

                                                      Last_Update  \
Country_Region                                                      
US              2020-04-01 21:58:492020-04-01 21:58:492020-04-...   

                        Lat          Long_  Confirmed  Deaths  Recovered  \
Country_Region                                                             
US              82956.96013 -197553.963757     216936    5222       8474   

                Active                                       Combined_Key  
Country_Region                                                             
US              209150  Abbeville, South Carolina, USAcadia, Louisiana...

## 6. 컬럼 타입 변경하기

#### **6.1** pandas에서 데이터 타입은 dtype 으로 불리우며, 주요 데이터 타입은 다음과 같음
- `object` 는 파이썬의 str 또는 혼용 데이터 타입 (문자열)
- `int64` 는 파이썬의 int (정수)
- `float64` 는 파이썬의 float (부동소숫점)
- `bool` 는 파이썬의 bool (True 또는 False 값을 가지는 boolean)

In [78]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc[['Country/Region', 'Confirmed']]
doc.info() # 데이터의 타입을 알아보기 위해

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country/Region  76 non-null     object 
 1   Confirmed       66 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.3+ KB


In [80]:
doc.head()

Country/Region  Confirmed
0  Mainland China        1.0
1  Mainland China       14.0
2  Mainland China        6.0
3  Mainland China        1.0
4  Mainland China        NaN

#### **6.2** `astype({컬럼명: 변경할타입})` : 특정 컬럼의 타입을 변경
- 변경할 데이터에 없는 데이터(NaN)이 있을 경우, 에러가 날 수 있음

In [83]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc[['Country/Region', 'Confirmed']] # 필요한 컬럼만 선택하기
doc = doc.dropna(subset=['Confirmed'])     # 특정 컬럼에 없는 데이터 삭제하기
doc = doc.astype({'Confirmed': 'int64'})   # 특정 컬럼의 데이터 타입 변경하기
doc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66 entries, 0 to 75
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Country/Region  66 non-null     object
 1   Confirmed       66 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.5+ KB


In [85]:
doc.head()

Country/Region  Confirmed
0  Mainland China          1
1  Mainland China         14
2  Mainland China          6
3  Mainland China          1
5  Mainland China         26

## 7. 데이터프레임 컬럼명 변경하기
- columns 로 컬럼명을 변경할 수 있음

In [88]:
doc = pd.read_csv(PATH + "01-22-2020.csv", encoding='utf-8-sig')
doc = doc[['Country/Region', 'Confirmed']] # 필요한 컬럼만 선택하기

In [90]:
doc.columns

Index(['Country/Region', 'Confirmed'], dtype='object')

In [94]:
doc.columns = ['Country_Region', 'Confirmed']
doc.columns

Index(['Country_Region', 'Confirmed'], dtype='object')

## 8. 데이터프레임에서 중복 행 확인/제거하기

#### **8.1** `duplicated()` : 중복 행 확인하기

In [100]:
doc = pd.read_csv("00_Material(Uploaded)/COVID-19-master/csse_covid_19_data/UID_ISO_FIPS_LookUp_Table.csv", encoding='utf-8-sig')
doc = doc[['iso2', 'Country_Region']]
doc

iso2 Country_Region
0      BW       Botswana
1      BI        Burundi
2      SL   Sierra Leone
3      AF    Afghanistan
4      AL        Albania
...   ...            ...
3555   US             US
3556   US             US
3557   US             US
3558   US             US
3559   US             US

[3560 rows x 2 columns]

In [102]:
doc.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
3555     True
3556     True
3557     True
3558     True
3559     True
Length: 3560, dtype: bool

#### **8.2** 중복된 행만 확인하기

In [105]:
doc[doc.duplicated()]

iso2  Country_Region
194    GB  United Kingdom
200    AU       Australia
201    AU       Australia
202    AU       Australia
203    AU       Australia
...   ...             ...
3555   US              US
3556   US              US
3557   US              US
3558   US              US
3559   US              US

[3353 rows x 2 columns]

#### **8.3** `drop_ducplicates()` : 중복 행 삭제중복값
- 특정 컬럼을 기준으로 중복 행 제거하기
    - `subset=특정컬럼`
- 중복된 경우, 처음과 마지막 행 중 어느 행을 남길 것인지 결정하기
    - 처음: `keep='first'` (디폴트)
    - 처음: keep='last'

In [108]:
doc = doc.drop_duplicates(subset='Country_Region', keep='last')
doc

iso2  Country_Region
0      BW        Botswana
1      BI         Burundi
2      SL    Sierra Leone
3      AF     Afghanistan
4      AL         Albania
...   ...             ...
198    TC  United Kingdom
206    AU       Australia
221    CA          Canada
254    CN           China
3559   US              US

[180 rows x 2 columns]